source/agora/utils/Test.d

```diff
@@ -627,60 +632,360 @@ unittest
-    Transaction result = Transaction(type);
-    foreach (addr; toward)
-        result.outputs ~= Output(amount, addr);
+    ***************************************************************************/
 
-    // Add the remainder to the first output.
-    result.outputs[0].value.mustAdd(remainder);
-    const inputHash = input.hashFull();
+    private KeyPair getOwnerKeyPair (scope const PublicKey address) @safe
+    {
+        // FIXME: Using the binary version didn't work (?)
+        if (address.toString() == "GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ")
+            return  () @trusted {
+                return KeyPair.fromSeed(Seed.fromString(
+                    "SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4"));
+            }();
+
+        return WK.Keys[address];
+    }
+
+    /***************************************************************************
+
+        Resets the state and changes the address of the refund transaction
+
+        Ideally one should provide the correct address for the refund
+        transaction in the constructor, however if for some reason it is not
+        known in advance, or if the `Output` or `Transaction` overload is used,
+        this function provides a convenient mean to change the refund address.
+
+        Params:
+            toward = Refund address to use for the transaction being built
+
+        Returns:
+            Reference to `this` for easy chaining
+
+    ***************************************************************************/
+
+    public ref typeof(this) refund (scope const PublicKey toward)
+        @safe return
+    {
+        assert(this.inputs.length > 0);
+
+        this.leftover = Output(Amount(0), toward);
+        this.inputs.each!(val => this.leftover.value.mustAdd(val.output.value));
+        this.data.outputs = null;
+
+        return this;
+    }
+
+    /***************************************************************************
+
+        Splits the attached input into multiple outputs of the given amounts.
+
+        Any leftover will remain in the refund transaction.
+        The value drawn (`amount * toward.length`) must be lesser or equal
+        to the fund currently attached and in the refund transaction.
 
-    // Add support for Transactions with multiple recipients
-    foreach (idx, const ref _; input.outputs)
-        result.inputs ~= Input(inputHash, cast(uint) idx);
+        Params:
+            amount = `Amount` to distribute to each `PublicKey`
+            toward = Array of `PublicKey` to give `amount` to
 
-    scope sign = (KeyPair kp, Hash h) @trusted { return kp.secret.sign(h[]); };
+        Returns:
+            Reference to `this` for easy chaining
 
-    const resultHash = result.hashFull();
-    foreach (idx, ref in_; result.inputs)
+    ***************************************************************************/
+
+    public ref typeof(this) draw (Amount amount, in PublicKey[] toward... )
```

Comment:
Possible future improvement: An overload that accepts a `const Amount[PublicKey]`

Reply:
Also range overload